features/frameworks/greentea-client/source/greentea_test_env.cpp

```diff
@@ -71,15 +72,25 @@ void _GREENTEA_SETUP_COMMON(const int timeout, const char *host_test_name, char
     // Sync preamble: "{{__sync;0dad4a9d-59a3-4aec-810d-d5fb09d852c1}}"
     // Example value of sync_uuid == "0dad4a9d-59a3-4aec-810d-d5fb09d852c1"
 
-    char _key[8] = {0};
+    char _key[11] = {0};
+    bool runTests = false;
 
-    while (1) {
+    while (!runTests) {
         greentea_parse_kv(_key, buffer, sizeof(_key), size);
-        greentea_write_string("mbedmbedmbedmbedmbedmbedmbedmbed\r\n");
         if (strcmp(_key, GREENTEA_TEST_ENV_SYNC) == 0) {
-            // Found correct __sync message
+            greentea_write_string("Host Key parsed\r\n");
+            // Send back the key to Host and wait for acknowledgment
             greentea_send_kv(_key, buffer);
-            break;
+        }
+        /**
+          *  Once the host key is parsed, we would send back that key to host and wait 
+          *  for an acknowledgment from host that the key we sent back is correct. If we dont
+          *  get back the ack, we keep waiting and parse values further until we get the ack. This
+          *  is to take care of issues when we encounter old keys in buffer and assume them to be right ones.
+          */
+        if (strcmp(_key, GREENTEA_TEST_HOST_ACK) == 0) {
```

Comment:
How this works in a case that never get the ACK?

Reply:
We sort of keep waiting until we get the ack. It's a blocking call, much like what we have right now for receiving the sync message.